Add 256 Color Support

jonathanabennett opened the pull request:
This commit addresses issue #2. It adds 256 color support very naively.
It does not check for terminal support of 256 colors. It does not check
that the int provided is in the range 0..255. It does not have any error
checking or handling of any sort.

Fixes: #2

jonathanabennett committed:
Add 256 Color Support

This commit addresses issue #2. It adds 256 color support very naively.
It does not check for terminal support of 256 colors. It does not check
that the int provided is in the range 0..255. It does not have any error
checking or handling of any sort.

Fixes: #2

jonathanabennett commented:
New PR with README commit removed.

dom96 commented:
Thank you.

Now I must ask: have you tested this?

If I'm honest I would prefer if this code took a ``Color`` type from the stdlib ``colors`` module but that would clash with the ``Color`` type defined by this module :(

Perhaps adding a little documentation comment on how to pass colours would be enough, ``0xFF1234`` right?

jonathanabennett commented:
Yes, I have tested this. It's currently handling the terrain colors in a roguelike I'm working on. I will keep hacking at this because I'm unhappy having to guess at color codes myself. It also gives me a chance to familiarize myself with a bit of Nim that I don't expect to use in my current code.

Termbox takes 256 color numbers, as seen [here](https://jonasjacek.github.io/colors/). The numbers that Termbox (and therefore nimbox) use are the color immediately beside the swatch. We should be able to make stdlib `colors` work by using explicit namespace, correct? I will write `examples/256colors.nim` as documentation tomorrow and submit it as an additional pull request.

jonathanabennett committed:
Example for 256color support

This commit adds a worked example of 256 color support.

jonathanabennett commented:
I've written a test. If you can test it on your machine and see if it works, that would be great. It's in the same format as the others (only displaying for 1 second before closing). I don't like that, but I kept it consistent with the style of the others.

dom96 commented on examples/colors256.nim:
```diff
@@ -0,0 +1,19 @@
+
+  nb.outputMode = out256
+
+#Colors are defined using xTerm numbers.
```
X term numbers? That's a bit weird. if there is 256 of these colours then I think an enum would be better

jonathanabennett replied:
I agree. I'm packing out my house this weekend, but should have time to go through and enumerate them all by Wednesday.

jonathanabennett replied:
OK, I've done some reading to find a list of color names and it looks like an enum is not the best option because the exact colors vary slightly from terminal emulator to terminal emulator (gotta love 1980s tech). I have instead used a Subrange, which restricts the user from inputting something outside the printable range.

jonathanabennett committed:
Subrange Restriction for 256 Colors print proc.

By using a SubRange called Colors256, we can ensure that the print
procedure is never passed a color it cannot print. Further reading
on the 256 colors palette suggests that fixing names for the colors
probably will not work as each terminal defines them slightly
diffferently.

dom96 commented:
Okay, this should be good. Thanks!
